### PR TITLE
python27: Support for Apple silicon

### PR DESCRIPTION
**Pull request by gmpalter**
#### Description

Add support for Apple silicon to python27:

-  Update configure to recognize "arm64" as the output of /usr/bin/arch
-  Update Mac/Tools/pythonw.c to handle the __arm64__ preprocessor definition

Fixes https://trac.macports.org/ticket/60968

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 11.0 20A5364e (Beta 6 on Apple silicon)
Xcode 12.0 12A8189n (Beta 6)

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Review by michaelld — APPROVED** [review]
works for me ; thanks!

**Comment by neverpanic** [comment]
Your commit message should contain "Closes: https://trac.macports.org/ticket/60968", since that will automatically close the issue and post a message to the ticket.

Also, your commit message should start with the name of the port you're modifying, so it should read "python27: Support Apple silicon"

**Comment by gmpalter** [comment]
I have updated the commit message.

**Comment by neverpanic** [comment]
There's an extra closing quote after the ticket link in the commit message, but I think our regex for ticket closing ignores that.

Let's give @jmroot a day or so to look at this, and if he doesn't respond by then we can still merge it as-is.

The CI seems to fail on 10.14 and 10.15 because gperf-bootstrap fails to link. I'm not familiar with this issue, but it does not seem to be caused by your change.

**Comment by kencu** [comment]
```gperf-bootstrap``` is hard-coded to use ```-stdlib=libstdc++```

**Comment by jmroot** [comment]
> Let's give @jmroot a day or so to look at this, and if he doesn't respond by then we can still merge it as-is.

Maintainer timeout is 3 days, please don't merge PRs for ports maintained by other people earlier than that. https://guide.macports.org/chunked/project.update-policies.html

**Comment by jmroot** [comment]
@gmpalter Thanks for the PR, however there are a number of other fixes needed, particularly in ctypes. I'll try to finish those soon.

**Review by ryandesign — CHANGES_REQUESTED** [review]
Yup there's *much* more to it than just this.

**Comment by michaelld** [comment]
OK ... so, there's _much_ more to be done ... but, this small change _does_ allow the python27 port to build on ARM64. In my initial testing, the "python2.7" executable works and runs Python scripts ... I haven't tested it exhaustively, so there's certainly room for "_much_ more to be done".

I'm all for getting even _basic_ support going for ARM64, as I've done with Boost. There very well might be "_much_ more to be done" for Boost, too, IDK. We have to start with supporting _somewhere_, and this PR provides that basic starting point for this port in my testing.

Thus for my curiosity, when y'all note that there is "_much_ more to be done" ... is it just for this port? for other related ports (e.g., dependencies)? how long are we talking about ... a few days? weeks? no idea because there is so _much_ more to be done?

If this extra work will take a week or longer, then my vote is to merge this PR noting that it provides _basic_ functionality with ARM64, and that more fixes are on the way. Incremental tweaks stand a chance of allowing folks using ARM64 -- not very many I'll admit -- to move on to other ports. Without these tweaks one can't even build python27 & hence can't move on to testing dependencies.

**Comment by ryandesign** [comment]
In this case I was referring specifically to python and the several pull requests that Apple engineers already submitted to the python project for both ARM and macOS 11 support which were merged there but which will of course not be backported to python27 since it is EOL. I have no idea what the consequences are of including only the small changes in this PR without the other changes.

**Comment by michaelld** [comment]
> In this case I was referring specifically to python and the several pull requests that Apple engineers already submitted to the python project for both ARM and macOS 11 support which were merged there but which will of course not be backported to python27 since it is EOL. I have no idea what the consequences are of including only the small changes in this PR without the other changes.

ACKnowledged ... I'll go hunt down those merges & see about getting them either into this PR or another. Thx for the pointer!

**Comment by jmroot** [comment]
I'm done backporting all but the ctypes changes. Those are the most complex. Continuing to work on them.

**Comment by michaelld** [comment]
> I'm done backporting all but the ctypes changes. Those are the most complex. Continuing to work on them.

Nice! I got pulled off onto other topics, so I'm glad you've found the time to do this! I'm happy to test once you get to that point.

**Comment by michaelld** [comment]
@jmroot any progress? I should have time this week to help backport changes if it makes sense for me to help out.

**Comment by jmroot** [comment]
I pushed the darwin 20 fixes in a46ff54. Still working on the arm64-specific bits.

**Comment by gmpalter** [comment]
This probably isn't the best place to mention this but I'm not sure where to bring it up.

Starting with macOS Big Sur beta 6, Apple silicon requires that all executables be signed. The linker generates adhoc signatures for any executable built with Xcode 12.0 beta 4 or later.

Tools like strip and install_name_tool will invalidate the signature. Lots of build scripts, including for python3.8 and, I assume, python2.7, do this. This requires resigning the executables.

I would open a ticket about this but it's a problem for multiple ports and I don't know how to do that.

Here's the relevant section of the macOS Big Sure Beta 7 for Apple silicon Release Notes.

> New in macOS 11 on Apple silicon Mac computers, and starting in macOS Big Sur 11 beta 6, the operating system enforces that any executable must be signed before it’s allowed to run. There isn’t a specific identity requirement for this signature: a simple ad-hoc signature is sufficient. This new behavior doesn’t change the long-established policy that our users and developers can run arbitrary code on their Macs, and is designed to simplify the execution policies on Apple silicon Mac computers and enable the system to better detect code modifications. This new policy doesn’t apply to translated x86 binaries running under Rosetta 2, nor does it apply to macOS 11 running on Intel-based platforms.
>
>To reduce the impact on existing development workflows, starting with Xcode 12 beta 4, the toolchain will now automatically sign your executables whenever you build from Xcode, or use command-line utilities such as clang(1) or ld(1). Since this signing mechanism generates signatures directly at link time, and doesn’t cover any resource other than the executable, it is expected to be faster than a traditional codesign(1) invocation. If you use a custom workflow involving tools that modify a binary after linking (e.g. strip or install_name_tool) you might need to manually call codesign(1) as an additional build phase to properly ad-hoc sign your binary. These new signatures are not bound to the specific machine that was used to build the executable, they can be verified on any other system and will be sufficient to comply with the new default code signing requirement on Apple silicon Mac computers. However, given that these signatures do not bear any valid identity, binaries signed this way cannot pass through Gatekeeper.
>
>Xcode toolchains have been updated in the latest Xcode beta to automatically apply ad-hoc signatures at link time to support projects that might not already be configured for signing. Install the most current Xcode beta to ensure you have up-to-date tooling support.
Software previously built on your Developer Transition Kit might need to be either re-built with new tools, or re-signed with codesign -s - —preserve-metadata=identifier,entitlements,flags,runtime -f <path to mach-O>, to resolve issues from using earlier releases of the toolchain, or correct executables that have incorrect or missing code signatures. To check the validity of a signature, run codesign -v -vvvv —ignore-resources <path to mach-O>. If the command indicates “valid on disk (not all contents verified)” and “satisfies its Designated Requirement”, the code is valid under this policy. (51911409)

**Comment by kencu** [comment]
For stuff like this, we usually turn to our resident guru, @jeremyhu , to help us see the way forward...

**Comment by saagarjha** [comment]
So, about "other problems": I had a patch for this a while back when @tobypeterson submitted a fix for python38: https://github.com/macports/macports-ports/commit/4a612d6f2f632c5b2d46b3d7a4ee9c623bab8e66. I held off on it because the tests fail; for more complicated programs that use libffi Python will segfault since it's not been made aware of Apple silicon's W^X. For now I have been working around it [with a somewhat crude hack](https://gist.github.com/saagarjha/d1ddd98537150e4a09520ed3ede54f5e) but at some point someone will have to fix libffi properly.

**Comment by michaelld** [comment]
FYI for folks looking for the `codesign` info, it's [here](https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11-universal-apps-beta-release-notes) ... this is a royal pain, by the way. I highly recommend folks with Apple Silicon not update to 11.0b7 just yet. Not all of MacPorts installs are broken, but enough are to make it painful and challenging. Sigh.

**Comment by gmpalter** [comment]
> FYI for folks looking for the `codesign` info, it's [here](https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11-universal-apps-beta-release-notes) ... this is a royal pain, by the way. I highly recommend folks with Apple Silicon not update to 11.0b7 just yet. Not all of MacPorts installs are broken, but enough are to make it painful and challenging. Sigh.

Indeed. For the record, I had to rebuild xorg and all its dependencies. When I was finished, there were somewhere around 400 executables that had to be manually resigned. Sigh

**Comment by kencu** [comment]
> there were somewhere around 400 executables that had to be manually resigned.

well that's insane...

**Comment by saagarjha** [comment]
I'm curious if we could insert our own `strip` and `install_name_tool` wrappers that sign the binary if the signature is broken, or perhaps do something with `port contents`.

**Comment by kencu** [comment]
exactly what I was thinking. A little script -- we already use such a script with our cctools +xcode variants anyway...

**Comment by gmpalter** [comment]
> > there were somewhere around 400 executables that had to be manually resigned.
> 
> well that's insane...

Indeed it is.  But, I used

> find . -type f -perm +a=x,g=x,o=x -exec codesign -v {} \; 2> /tmp/foo; grep invalid /tmp/foo > /tmp/fixup.sh; rm /tmp/foo

to find the broken executables. An emacs macro to massage the output and everything's re-signed. (I'm sure I could've piped that through sed instead to make the script in one shot.)

**Comment by michaelld** [comment]
Nice! I've had no luck with the re-signing process. Really not sure what I'm not doing correctly!

I love the idea of wrappers for those executables; that would be a nice and easy way to handle the challenging IMHO!

**Comment by saagarjha** [comment]
The issue with using execute permissions to detect these is that they won't work for shared libraries that are missing those permissions, and they will try to codesign shell scripts and other files that *are* marked with +x but do not need a signature.

**Comment by gmpalter** [comment]
> The issue with using execute permissions to detect these is that they won't work for shared libraries that are missing those permissions, and they will try to codesign shell scripts and other files that _are_ marked with +x but do not need a signature.

codesign doesn't generate an error with "invalid" for shell scripts and other files that are marked executable. (The error it generates is "code object is not signed at all".)

If a shared library doesn't have execute permission, I would presume it can't be executed so the signature doesn't come into play. (Please correct me if I'm wrong about that.)

**Comment by saagarjha** [comment]
> codesign doesn't generate an error with "invalid" for shell scripts and other files that are marked executable. (The error it generates is "code object is not signed at all".)

Right, but relying on an error case here seems kind of unclean…

> If a shared library doesn't have execute permission, I would presume it can't be executed so the signature doesn't come into play. (Please correct me if I'm wrong about that.)

You cannot execute it as in you can't literally go "./libfoo.dylib" anymore (not that you were likely doing this previously), but you can still load it into a process and at that point I believe the lack of a signature will become a problem.

**Comment by michaelld** [comment]
I'm trying to get PythonX.Y working ... for example 2.7 or 3.7. So I go find the actual executable and the codesign check fails for it. So, I fix that & now codesign looks valid. I check the dependency libraries & follow those, checking each in turn; all come back as valid signed. But, when I try to execute pythonX.Y, it crashes with a bad signature. The executable & libraries still pass codesign verification ... no idea what's going on. @gmpalter have you has success with getting Python working? Any hints?

**Comment by kencu** [comment]
Here's the bit in our cctools port that writes the script. https://github.com/macports/macports-ports/blob/1505aec2fc08e2d08066f854ff594f3aa63cb273/devel/cctools/Portfile#L116

Probably only needs one more line added for the codesign instruction -- although perhaps only for certain tools...

**Comment by jmroot** [comment]
> I'm curious if we could insert our own `strip` and `install_name_tool` wrappers that sign the binary if the signature is broken, or perhaps do something with `port contents`.

A new step in destroot_finish would certainly be possible. But can we move this discussion to somewhere more appropriate, like the mailing list or a Trac ticket?

**Comment by michaelld** [comment]
yeah that would be best ... this PR is closed & though the conversation is related, it's not about the PR itself

**Comment by kencu** [comment]
Why is install_name_tool skipped there -- hmm. that needs fixing.

**Comment by gmpalter** [comment]
I subscribed to the macports-dev mailing list but haven't yet been approved. I can't send reply to the thread on the code signing issue. So, I'm posting this here in the hope that someone will pass it along.

Apple today released Xcode 12.2 beta 2 and its Release Notes state

>### Apple Clang Compiler
>### Resolved Issues
>Fixed an issue that caused strip, install_name_tool and vtool to corrupt the ad-hoc code signatures generated by the linker for arm64 Mach-O files. (51911417)

**Comment by michaelld** [comment]
Sorry to folks for the short continued discussion on this PR!

My advice is to subscribe to the MP users list instead & go with that instead of MP dev. There's a lot of overlap anyway.

Thanks for the tip on the new Xcode 12.2 beta 2 .. I'm downloading right now & will give this new Xcode a go starting later today ... 11.25 GB ... ¿why oh why? 1+ hour download for me ... sigh ...

**Pull request closed without merging**
